### PR TITLE
Ports xenomedes' fusion rework

### PR DESCRIPTION
**Pull request by jupyterkat**
from base BeeStation/BeeStation-Hornet#2614

Dm side changes:
```
/proc/fusion_ball(datum/holder, reaction_energy, standard_energy)
	var/turf/open/location
	if (istype(holder,/datum/pipeline)) //Find the tile the reaction is occuring on, or a random part of the network if it's a pipenet.
		var/datum/pipeline/fusion_pipenet = holder
		location = get_turf(pick(fusion_pipenet.members))
	else
		location = get_turf(holder)
	if(location)
		if(prob(PERCENT(((PARTICLE_CHANCE_CONSTANT)/(reaction_energy-PARTICLE_CHANCE_CONSTANT)) + 1))) //Asymptopically approaches 100% as the energy of the reaction goes up.
			location.fire_nuclear_particle(customize = TRUE, custompower = standard_energy)
		radiation_pulse(location, max(2000 * 3 ** (log(10,standard_energy) - FUSION_RAD_MIDPOINT), 0))

```

### CODE REVIEW

File: src/reaction/hooks.rs [hunk 1]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
-					&Value::from(reaction_energy),
-					&Value::from(instability),
-				])?;
-			} else {
-				Proc::find(byond_string!("/proc/stack_trace"))
-					.ok_or_else(|| runtime!("Couldn't find stack_trace!"))?
-					.call(&[&Value::from_string(
-						"fusion_ball not found! Auxmos hooked fusion does not work without it!",
-					)?])?;
+			if new_heat_cap > MINIMUM_HEAT_CAPACITY{
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:293]:
```suggestion
			if new_heat_cap > MINIMUM_HEAT_CAPACITY {
```

File: src/reaction/hooks.rs [hunk 2]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
-		let delta_plasma = initial_plasma - plasma;
-		if delta_plasma < 0.0 {
-			if instability < FUSION_INSTABILITY_ENDOTHERMALITY {
-				0.0
-			} else {
-				delta_plasma
-					* PLASMA_BINDING_ENERGY
-					* (instability - FUSION_INSTABILITY_ENDOTHERMALITY).sqrt()
-			}
+		if (delta_plasma > 0.0) || (instability <= FUSION_INSTABILITY_ENDOTHERMALITY) {
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:242]:
```suggestion
		if delta_plasma > 0.0 || instability <= FUSION_INSTABILITY_ENDOTHERMALITY {
```

File: src/reaction/hooks.rs [hunk 3]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
+			(delta_plasma * PLASMA_BINDING_ENERGY)
+			.max(0.0)
 		} else {
 			delta_plasma * PLASMA_BINDING_ENERGY
+				* ((instability-FUSION_INSTABILITY_ENDOTHERMALITY).sqrt())
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:247]:
```suggestion
				* (instability - FUSION_INSTABILITY_ENDOTHERMALITY).sqrt()
```

File: src/reaction/hooks.rs [hunk 4]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
 			))
 		})?;
-	let instability = (gas_power * INSTABILITY_GAS_FACTOR)
-		.powi(2)
+	//The size of the phase space hypertorus
+	let toroidal_size = TOROID_CALCULATED_THRESHOLD
+	+ { if temperature_scale <= FUSION_BASE_TEMPSCALE {
+		(temperature_scale - FUSION_BASE_TEMPSCALE) / FUSION_BUFFER_DIVISOR
+	} else {
+		(4.0_f32.powf(temperature_scale - FUSION_BASE_TEMPSCALE)) / FUSION_SLOPE_DIVISOR
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:215]:
```suggestion
		4.0_f32.powf(temperature_scale - FUSION_BASE_TEMPSCALE) / FUSION_SLOPE_DIVISOR
```

File: src/reaction/hooks.rs [hunk 5]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
+	if reaction_energy != 0.0 {
+		let middle_energy = (((TOROID_CALCULATED_THRESHOLD / 2.0) * scale_factor)
+			+ FUSION_MOLE_THRESHOLD)
+			* (200.0 * FUSION_MIDDLE_ENERGY_REFERENCE);
+		thermal_energy = middle_energy
+			* FUSION_ENERGY_TRANSLATION_EXPONENT
+			.powf((thermal_energy / middle_energy).log10());
+		//This bowdlerization is a double-edged sword. Tread with care!
+		let bowdlerized_reaction_energy = reaction_energy.clamp(
+			thermal_energy * ((1.0 / (FUSION_ENERGY_TRANSLATION_EXPONENT.powi(2))) - 1.0),
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:261]:
```suggestion
			thermal_energy * ((1.0 / FUSION_ENERGY_TRANSLATION_EXPONENT.powi(2)) - 1.0),
```

File: src/reaction/hooks.rs [hunk 6]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
+			thermal_energy * (FUSION_ENERGY_TRANSLATION_EXPONENT.powi(2) - 1.0));
+		thermal_energy = middle_energy
+			* 10_f32.powf(((thermal_energy + bowdlerized_reaction_energy)
+			/ middle_energy).log(FUSION_ENERGY_TRANSLATION_EXPONENT))
+	};
+
+	//The decay of the tritium and the reaction's energy produces waste gases, different ones depending on whether the reaction is endo or exothermic
+	let standard_waste_gas_output = scale_factor
+		* (FUSION_TRITIUM_CONVERSION_COEFFICIENT*FUSION_TRITIUM_MOLES_USED);
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:270]:
```suggestion
		* (FUSION_TRITIUM_CONVERSION_COEFFICIENT * FUSION_TRITIUM_MOLES_USED);
```

File: src/reaction/hooks.rs [hunk 7]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
-					&Value::from(instability),
-				])?;
-			} else {
-				Proc::find(byond_string!("/proc/stack_trace"))
-					.ok_or_else(|| runtime!("Couldn't find stack_trace!"))?
-					.call(&[&Value::from_string(
-						"fusion_ball not found! Auxmos hooked fusion does not work without it!",
-					)?])?;
+			if new_heat_cap > MINIMUM_HEAT_CAPACITY{
+				air.set_temperature((thermal_energy/new_heat_cap).clamp(TCMB, INFINITY));
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:294]:
```suggestion
				air.set_temperature((thermal_energy / new_heat_cap).clamp(TCMB, INFINITY));
```

File: src/reaction/hooks.rs [hunk 8]
```diff
@@ -174,109 +174,148 @@ fn tritfire(byond_air: &Value, holder: &Value) {
...
+			}
+		} else if reaction_energy == 0.0
+			&& instability <= FUSION_INSTABILITY_ENDOTHERMALITY {
+			if new_heat_cap > MINIMUM_HEAT_CAPACITY{
+				air.set_temperature((thermal_energy/new_heat_cap).clamp(TCMB, INFINITY)); //THIS SHOULD STAY OR FUSION WILL EAT YOUR FACE
```
Comment (BrumalLake) [inline src/reaction/hooks.rs:299]:
```suggestion
				air.set_temperature((thermal_energy / new_heat_cap).clamp(TCMB, INFINITY)); //THIS SHOULD STAY OR FUSION WILL EAT YOUR FACE
```

**Comment by Putnam3145** [comment]
note: this was actually merged with above

**Pull request closed without merging**